File: lib/ts/src/internal/schema/ParseMsgPactSchema.ts
import { MsgPactSchema } from '../../MsgPactSchema';
import { SchemaParseFailure } from './SchemaParseFailure';
import { VType } from '../types/VType';
import { VFieldDeclaration } from '../types/VFieldDeclaration';
import { MsgPactSchemaParseError } from '../../MsgPactSchemaParseError';
import { applyErrorToParsedTypes } from './ApplyErrorToParsedTypes';
import { catchErrorCollisions } from './CatchErrorCollisions';
import { findMatchingSchemaKey } from './FindMatchingSchemaKey';
import { findSchemaKey } from './FindSchemaKey';
import { getOrParseType } from './GetOrParseType';
import { getTypeUnexpectedParseFailure } from './GetTypeUnexpectedParseFailure';
import { parseErrorType } from './ParseErrorType';
import { parseHeadersType } from './ParseHeadersType';
import { VError } from '../types/VError';
import { ParseContext } from './ParseContext';
import { getPathDocumentCoordinatesPseudoJson } from './GetPathDocumentCoordinatesPseudoJson';
import { VHeaders } from '../types/VHeaders';
import { catchHeaderCollisions } from './CatchHeaderCollisions';

export function parseMsgPactSchema(msgPactSchemaDocumentNamesToJson: Record<string, string>): MsgPactSchema {
    const originalSchema: { [key: string]: Record<string, object> } = {};
    const fullSchema: { [key: string]: Record<string, object> } = {};
    const parsedTypes: { [key: string]: VType } = {};
    const parseFailures: SchemaParseFailure[] = [];
    const failedTypes: Set<string> = new Set();
    const schemaKeysToIndex: { [key: string]: number } = {};
    const schemaKeysToDocumentName: { [key: string]: string } = {};
    const schemaKeys: Set<string> = new Set();

    const orderedDocumentNames = Object.keys(msgPactSchemaDocumentNamesToJson).sort();

    const msgPactSchemaDocumentNamesToPseudoJson: Record<string, any[]> = {};

    for (const [documentName, jsonValue] of Object.entries(msgPactSchemaDocumentNamesToJson)) {
        let msgPactSchemaPseudoJsonInit: any;
        try {
            msgPactSchemaPseudoJsonInit = JSON.parse(jsonValue);
        } catch (e) {
            throw new MsgPactSchemaParseError(
                [new SchemaParseFailure(documentName, [], 'JsonInvalid', {})],
                msgPactSchemaDocumentNamesToJson,
                e as Error,
            );
        }

        if (!Array.isArray(msgPactSchemaPseudoJsonInit)) {
            const thisParseFailure = getTypeUnexpectedParseFailure(documentName, [], msgPactSchemaPseudoJsonInit, 'Array');
            throw new MsgPactSchemaParseError(thisParseFailure, msgPactSchemaDocumentNamesToJson);
        }
        const msgPactSchemaPseudoJson: any[] = msgPactSchemaPseudoJsonInit;

        msgPactSchemaDocumentNamesToPseudoJson[documentName] = msgPactSchemaPseudoJson;
    }

    for (const documentName of orderedDocumentNames) {
        const msgPactSchemaPseudoJson = msgPactSchemaDocumentNamesToPseudoJson[documentName];

        let index = -1;
        for (const definition of msgPactSchemaPseudoJson) {
            index += 1;
            const loopPath = [index];

            if (
                typeof definition !== 'object' ||
                Array.isArray(definition) ||
                definition === null ||
                definition === undefined
            ) {
                const thisParseFailures = getTypeUnexpectedParseFailure(
                    documentName,
                    loopPath as any[],
                    definition,
                    'Object',
                );
                parseFailures.push(...thisParseFailures);
                continue;
            }

            const def_: Record<string, object> = definition as Record<string, object>;

            try {
                const schemaKey = findSchemaKey(documentName, def_, index, msgPactSchemaDocumentNamesToJson);
                const matchingSchemaKey = findMatchingSchemaKey(schemaKeys, schemaKey);
                if (matchingSchemaKey !== null) {
                    const otherPathIndex = schemaKeysToIndex[matchingSchemaKey];
                    const otherDocumentName = schemaKeysToDocumentName[matchingSchemaKey];
                    const finalPath = [...loopPath, schemaKey];
                    const otherFinalPath = [otherPathIndex, matchingSchemaKey];
                    const otherDocumentJson = msgPactSchemaDocumentNamesToJson[otherDocumentName];
                    const otherLocationPseudoJson = getPathDocumentCoordinatesPseudoJson(
                        otherFinalPath,
                        otherDocumentJson,
                    );
                    parseFailures.push(
                        new SchemaParseFailure(documentName, finalPath as any[], 'PathCollision', {
                            document: otherDocumentName,
                            location: otherLocationPseudoJson,
                            path: [otherPathIndex, matchingSchemaKey],
                        }),
                    );
                    continue;
                }

                schemaKeys.add(schemaKey);
                schemaKeysToIndex[schemaKey] = index;
                schemaKeysToDocumentName[schemaKey] = documentName;
                if ('auto_' === documentName || !documentName.endsWith('_')) {
                    originalSchema[schemaKey] = def_;
                }
                fullSchema[schemaKey] = def_;
            } catch (e) {
                if (e instanceof MsgPactSchemaParseError) {
                    parseFailures.push(...e.schemaParseFailures);
                } else {
                    throw e;
                }
            }
        }
    }

    if (parseFailures.length > 0) {
        throw new MsgPactSchemaParseError(parseFailures, msgPactSchemaDocumentNamesToJson);
    }

    const headerKeys: Set<string> = new Set();
    const errorKeys: Set<string> = new Set();

    for (const schemaKey of schemaKeys) {
        if (schemaKey.startsWith('info.')) {
            continue;
        } else if (schemaKey.startsWith('headers.')) {
            headerKeys.add(schemaKey);
            continue;
        } else if (schemaKey.startsWith('errors.')) {
            errorKeys.add(schemaKey);
            continue;
        }

        const thisIndex = schemaKeysToIndex[schemaKey];
        const thisDocumentName = schemaKeysToDocumentName[schemaKey];

        try {
            getOrParseType(
                [thisIndex],
                schemaKey,
                new ParseContext(
                    thisDocumentName,
                    msgPactSchemaDocumentNamesToPseudoJson,
                    msgPactSchemaDocumentNamesToJson,
                    schemaKeysToDocumentName,
                    schemaKeysToIndex,
                    parsedTypes,
                    parseFailures,
                    failedTypes,
                ),
            );
        } catch (e) {
            if (e instanceof MsgPactSchemaParseError) {
                parseFailures.push(...e.schemaParseFailures);
            } else {
                throw e;
            }
        }
    }

    if (parseFailures.length > 0) {
        throw new MsgPactSchemaParseError(parseFailures, msgPactSchemaDocumentNamesToJson);
    }

    const errors: VError[] = [];

    try {
        for (const thisKey of errorKeys) {
            const thisIndex = schemaKeysToIndex[thisKey];
            const thisDocumentName = schemaKeysToDocumentName[thisKey];
            const msgPactSchemaPseudoJson = msgPactSchemaDocumentNamesToPseudoJson[thisDocumentName];
            const def_ = msgPactSchemaPseudoJson[thisIndex] as { [key: string]: any };

            try {
                const error = parseErrorType(
                    [thisIndex],
                    def_,
                    thisKey,
                    new ParseContext(
                        thisDocumentName,
                        msgPactSchemaDocumentNamesToPseudoJson,
                        msgPactSchemaDocumentNamesToJson,
                        schemaKeysToDocumentName,
                        schemaKeysToIndex,
                        parsedTypes,
                        parseFailures,
                        failedTypes,
                    ),
                );
                errors.push(error);
            } catch (e) {
                if (e instanceof MsgPactSchemaParseError) {
                    parseFailures.push(...e.schemaParseFailures);
                } else {
                    throw e;
                }
            }
        }
    } catch (e) {
        if (e instanceof MsgPactSchemaParseError) {
            parseFailures.push(...e.schemaParseFailures);
        } else {
            throw e;
        }
    }

    if (parseFailures.length > 0) {
        throw new MsgPactSchemaParseError(parseFailures, msgPactSchemaDocumentNamesToJson);
    }

    try {
        catchErrorCollisions(
            msgPactSchemaDocumentNamesToPseudoJson,
            errorKeys,
            schemaKeysToIndex,
            schemaKeysToDocumentName,
            msgPactSchemaDocumentNamesToJson,
        );
    } catch (e) {
        if (e instanceof MsgPactSchemaParseError) {
            parseFailures.push(...e.schemaParseFailures);
        } else {
            throw e;
        }
    }

    if (parseFailures.length > 0) {
        throw new MsgPactSchemaParseError(parseFailures, msgPactSchemaDocumentNamesToJson);
    }

    for (const error of errors) {
        try {
            applyErrorToParsedTypes(
                error,
                parsedTypes,
                schemaKeysToDocumentName,
                schemaKeysToIndex,
                msgPactSchemaDocumentNamesToJson,
            );
        } catch (e) {
            if (e instanceof MsgPactSchemaParseError) {
                parseFailures.push(...e.schemaParseFailures);
            } else {
                throw e;
            }
        }
    }

    const headers: VHeaders[] = [];

    for (const headerKey of headerKeys) {
        const thisIndex = schemaKeysToIndex[headerKey];
        const thisDocumentName = schemaKeysToDocumentName[headerKey];
        const msgPactSchemaPseudoJson = msgPactSchemaDocumentNamesToPseudoJson[thisDocumentName];
        const def_ = msgPactSchemaPseudoJson[thisIndex] as { [key: string]: any };

        try {
            const headerType = parseHeadersType(
                [thisIndex],
                def_,
                headerKey,
                new ParseContext(
                    thisDocumentName,
                    msgPactSchemaDocumentNamesToPseudoJson,
                    msgPactSchemaDocumentNamesToJson,
                    schemaKeysToDocumentName,
                    schemaKeysToIndex,
                    parsedTypes,
                    parseFailures,
                    failedTypes,
                ),
            );
            headers.push(headerType);
        } catch (e) {
            if (e instanceof MsgPactSchemaParseError) {
                parseFailures.push(...e.schemaParseFailures);
            } else {
                throw e;
            }
        }
    }

    if (parseFailures.length > 0) {
        throw new MsgPactSchemaParseError(parseFailures, msgPactSchemaDocumentNamesToJson);
    }

    try {
        catchHeaderCollisions(
            msgPactSchemaDocumentNamesToPseudoJson,
            headerKeys,
            schemaKeysToIndex,
            schemaKeysToDocumentName,
            msgPactSchemaDocumentNamesToJson,
        );
    } catch (e) {
        if (e instanceof MsgPactSchemaParseError) {
            parseFailures.push(...e.schemaParseFailures);
        } else {
            throw e;
        }
    }

    if (parseFailures.length > 0) {
        throw new MsgPactSchemaParseError(parseFailures, msgPactSchemaDocumentNamesToJson);
    }

    const requestHeaders: { [key: string]: VFieldDeclaration } = {};
    const responseHeaders: { [key: string]: VFieldDeclaration } = {};

    for (const header of headers) {
        Object.assign(requestHeaders, header.requestHeaders);
        Object.assign(responseHeaders, header.responseHeaders);
    }

    const sortKeys = (a: string, b: string) => {
        const aStartsWithInfo = a.startsWith('info.');
        const bStartsWithInfo = b.startsWith('info.');
        if (aStartsWithInfo && !bStartsWithInfo) return -1;
        if (!aStartsWithInfo && bStartsWithInfo) return 1;
        return a < b ? -1 : a > b ? 1 : 0;
    };

    const sortedSchemaKeys = Object.keys(originalSchema).sort(sortKeys);
    const finalOriginalSchema: Record<string, object>[] = sortedSchemaKeys.map((key) => originalSchema[key]);

    const sortedFullSchemaKeys = Object.keys(fullSchema).sort(sortKeys);
    const finalFullSchema: Record<string, object>[] = sortedFullSchemaKeys.map((key) => fullSchema[key]);

    return new MsgPactSchema(finalOriginalSchema, finalFullSchema, parsedTypes, requestHeaders, responseHeaders);
}

File: lib/java/src/main/java/io/github/msgpact/internal/schema/ParseMsgPactSchema.java
package io.github.msgpact.internal.schema;

import static io.github.msgpact.internal.schema.ApplyErrorToParsedTypes.applyErrorToParsedTypes;
import static io.github.msgpact.internal.schema.CatchErrorCollisions.catchErrorCollisions;
import static io.github.msgpact.internal.schema.CatchHeaderCollisions.catchHeaderCollisions;
import static io.github.msgpact.internal.schema.FindMatchingSchemaKey.findMatchingSchemaKey;
import static io.github.msgpact.internal.schema.FindSchemaKey.findSchemaKey;
import static io.github.msgpact.internal.schema.GetOrParseType.getOrParseType;
import static io.github.msgpact.internal.schema.GetTypeUnexpectedParseFailure.getTypeUnexpectedParseFailure;
import static io.github.msgpact.internal.schema.ParseErrorType.parseErrorType;
import static io.github.msgpact.internal.schema.ParseHeadersType.parseHeadersType;

import java.io.IOException;
import java.util.*;

import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;

import io.github.msgpact.MsgPactSchema;
import io.github.msgpact.MsgPactSchemaParseError;
import io.github.msgpact.internal.types.VError;
import io.github.msgpact.internal.types.VFieldDeclaration;
import io.github.msgpact.internal.types.VHeaders;
import io.github.msgpact.internal.types.VType;

public class ParseMsgPactSchema {

    public static MsgPactSchema parseMsgPactSchema(Map<String, String> msgPactSchemaNameToJson) {
        final var originalSchema = new TreeMap<String, Object>((k1, k2) -> {
            if (k1.equals(k2)) {
                return 0;
            } else if (k1.startsWith("info.")) {
                return -1;
            } else if (k2.startsWith("info.")) {
                return 1;
            } else {
                return k1.compareTo(k2);
            }
        });
        final var parsedTypes = new HashMap<String, VType>();
        final var parseFailures = new ArrayList<SchemaParseFailure>();
        final var failedTypes = new HashSet<String>();
        final var schemaKeysToIndex = new HashMap<String, Integer>();
        final var schemaKeysToDocumentName = new HashMap<String, String>();
        final var schemaKeys = new HashSet<String>();

        final var orderedDocuments = new TreeSet<>(msgPactSchemaNameToJson.keySet());

        final var objectMapper = new ObjectMapper();
        final var msgPactSchemaNameToPseudoJson = new HashMap<String, List<Object>>();

        for (var msgPactSchemaJson : msgPactSchemaNameToJson.entrySet()) {
            var documentName = msgPactSchemaJson.getKey();

            final Object msgPactSchemaPseudoJsonInit;
            try {
                msgPactSchemaPseudoJsonInit = objectMapper.readValue(msgPactSchemaJson.getValue(),
                        new TypeReference<>() {
                        });
            } catch (IOException e) {
                throw new MsgPactSchemaParseError(
                        List.of(new SchemaParseFailure(documentName, List.of(), "JsonInvalid", Map.of())),
                        msgPactSchemaNameToJson,
                        e);
            }

            if (!(msgPactSchemaPseudoJsonInit instanceof List)) {
                final List<SchemaParseFailure> thisParseFailure = getTypeUnexpectedParseFailure(documentName, List.of(),
                        msgPactSchemaPseudoJsonInit, "Array");
                throw new MsgPactSchemaParseError(thisParseFailure, msgPactSchemaNameToJson);
            }
            final List<Object> msgPactSchemaPseudoJson = (List<Object>) msgPactSchemaPseudoJsonInit;

            msgPactSchemaNameToPseudoJson.put(documentName, msgPactSchemaPseudoJson);
        }

        for (var documentName : orderedDocuments) {
            var msgPactSchemaPseudoJson = msgPactSchemaNameToPseudoJson.get(documentName);

            int index = -1;
            for (Object definition : msgPactSchemaPseudoJson) {
                index++;
                var loopPath = new ArrayList<Object>(List.of(index));

                if (!(definition instanceof Map)) {
                    var thisParseFailures = getTypeUnexpectedParseFailure(
                            documentName, (List<Object>) loopPath, definition, "Object");
                    parseFailures.addAll(thisParseFailures);
                    continue;
                }

                var def = (Map<String, Object>) definition;

                try {
                    var schemaKey = findSchemaKey(documentName, def, index, msgPactSchemaNameToJson);
                    var matchingSchemaKey = findMatchingSchemaKey(schemaKeys, schemaKey);
                    if (matchingSchemaKey != null) {
                        var otherPathIndex = schemaKeysToIndex.get(matchingSchemaKey);
                        var otherDocumentName = schemaKeysToDocumentName.get(matchingSchemaKey);
                        var finalPath = new ArrayList<>(loopPath);
                        finalPath.add(schemaKey);
                        List<Object> finalOtherPath = new ArrayList<>(List.of(otherPathIndex, matchingSchemaKey));
                        var finalOtherDocumentJson = msgPactSchemaNameToJson.get(otherDocumentName);
                        var finalOtherLocationPseudoJson = GetPathDocumentCoordinatesPseudoJson
                                .getPathDocumentCoordinatesPseudoJson(finalOtherPath, finalOtherDocumentJson);

                        parseFailures.add(new SchemaParseFailure(
                                documentName,
                                finalPath,
                                "PathCollision",
                                Map.of("document", otherDocumentName, "path", finalOtherPath,
                                        "location", finalOtherLocationPseudoJson)));
                        continue;
                    }

                    schemaKeys.add(schemaKey);
                    schemaKeysToIndex.put(schemaKey, index);
                    schemaKeysToDocumentName.put(schemaKey, documentName);
                    if ("auto_".equals(documentName) || !documentName.endsWith("_")) {
                        originalSchema.put(schemaKey, def);
                    }
                } catch (MsgPactSchemaParseError e) {
                    parseFailures.addAll(e.schemaParseFailures);
                }
            }
        }

        if (!parseFailures.isEmpty()) {
            throw new MsgPactSchemaParseError(parseFailures, msgPactSchemaNameToJson);
        }

        var headerKeys = new HashSet<String>();
        var responseHeaderKeys = new HashSet<String>();
        var errorKeys = new HashSet<String>();

        for (String schemaKey : schemaKeys) {
            if (schemaKey.startsWith("info.")) {
                continue;
            } else if (schemaKey.startsWith("headers.")) {
                headerKeys.add(schemaKey);
                continue;
            } else if (schemaKey.startsWith("errors.")) {
                errorKeys.add(schemaKey);
                continue;
            }

            var thisIndex = schemaKeysToIndex.get(schemaKey);
            var thisDocumentName = schemaKeysToDocumentName.get(schemaKey);

            try {
                getOrParseType(
                        List.of(thisIndex),
                        schemaKey,
                        new ParseContext(
                                thisDocumentName,
                                msgPactSchemaNameToPseudoJson,
                                msgPactSchemaNameToJson,
                                schemaKeysToDocumentName,
                                schemaKeysToIndex,
                                parsedTypes,
                                parseFailures,
                                failedTypes));
            } catch (MsgPactSchemaParseError e) {
                parseFailures.addAll(e.schemaParseFailures);
            }
        }

        if (!parseFailures.isEmpty()) {
            throw new MsgPactSchemaParseError(parseFailures, msgPactSchemaNameToJson);
        }

        var errors = new ArrayList<VError>();

        for (String thisKey : errorKeys) {
            var thisIndex = schemaKeysToIndex.get(thisKey);
            var thisDocumentName = schemaKeysToDocumentName.get(thisKey);
            var msgPactSchemaPseudoJson = msgPactSchemaNameToPseudoJson.get(thisDocumentName);
            var def = (Map<String, Object>) msgPactSchemaPseudoJson.get(thisIndex);

            try {
                var error = parseErrorType(
                        List.of(thisIndex),
                        def,
                        thisKey,
                        new ParseContext(
                                thisDocumentName,
                                msgPactSchemaNameToPseudoJson,
                                msgPactSchemaNameToJson,
                                schemaKeysToDocumentName,
                                schemaKeysToIndex,
                                parsedTypes,
                                parseFailures,
                                failedTypes));
                errors.add(error);
            } catch (MsgPactSchemaParseError e) {
                parseFailures.addAll(e.schemaParseFailures);
            }
        }

        if (!parseFailures.isEmpty()) {
            throw new MsgPactSchemaParseError(parseFailures, msgPactSchemaNameToJson);
        }

        try {
            catchErrorCollisions(msgPactSchemaNameToPseudoJson, errorKeys, schemaKeysToIndex, schemaKeysToDocumentName,
                    msgPactSchemaNameToJson);
        } catch (MsgPactSchemaParseError e) {
            parseFailures.addAll(e.schemaParseFailures);
        }

        if (!parseFailures.isEmpty()) {
            throw new MsgPactSchemaParseError(parseFailures, msgPactSchemaNameToJson);
        }

        for (var error : errors) {
            try {
                applyErrorToParsedTypes(error, parsedTypes, schemaKeysToDocumentName, schemaKeysToIndex,
                        msgPactSchemaNameToJson);
            } catch (MsgPactSchemaParseError e) {
                parseFailures.addAll(e.schemaParseFailures);
            }
        }

        var headers = new ArrayList<VHeaders>();

        var requestHeaders = new HashMap<String, VFieldDeclaration>();
        var responseHeaders = new HashMap<String, VFieldDeclaration>();

        for (String headerKey : headerKeys) {
            var thisIndex = schemaKeysToIndex.get(headerKey);
            var thisDocumentName = schemaKeysToDocumentName.get(headerKey);
            var msgPactSchemaPseudoJson = msgPactSchemaNameToPseudoJson.get(thisDocumentName);
            var def = (Map<String, Object>) msgPactSchemaPseudoJson.get(thisIndex);

            try {
                var headerType = parseHeadersType(
                        List.of(thisIndex),
                        def,
                        headerKey,
                        new ParseContext(
                                thisDocumentName,
                                msgPactSchemaNameToPseudoJson,
                                msgPactSchemaNameToJson,
                                schemaKeysToDocumentName,
                                schemaKeysToIndex,
                                parsedTypes,
                                parseFailures,
                                failedTypes));
                headers.add(headerType);
            } catch (MsgPactSchemaParseError e) {
                parseFailures.addAll(e.schemaParseFailures);
            }
        }

        if (!parseFailures.isEmpty()) {
            throw new MsgPactSchemaParseError(parseFailures, msgPactSchemaNameToJson);
        }

        try {
            catchHeaderCollisions(
                    msgPactSchemaNameToPseudoJson, headerKeys, schemaKeysToIndex, schemaKeysToDocumentName,
                    msgPactSchemaNameToJson);
        } catch (MsgPactSchemaParseError e) {
            parseFailures.addAll(e.schemaParseFailures);
        }

        if (!parseFailures.isEmpty()) {
            throw new MsgPactSchemaParseError(parseFailures, msgPactSchemaNameToJson);
        }

        for (final var header : headers) {
            requestHeaders.putAll(header.requestHeaders);
            responseHeaders.putAll(header.responseHeaders);

        }

        if (!parseFailures.isEmpty()) {
            throw new MsgPactSchemaParseError(parseFailures, msgPactSchemaNameToJson);
        }

        final var finalOriginalSchema = new ArrayList<>(originalSchema.values());

        return new MsgPactSchema(
                finalOriginalSchema,
                parsedTypes,
                requestHeaders,
                responseHeaders);
    }
}
